File: android/src/main/java/com/lightningkite/khrysalis/location/GeoAddress.actual.kt
package com.lightningkite.khrysalis.location

import android.location.Address
import com.lightningkite.khrysalis.location.GeoCoordinate
import java.util.*

typealias GeoAddress = Address

/* SHARED DECLARATIONS
val GeoAddress.locality: String
val GeoAddress.adminArea: String
val GeoAddress.countryName: String
 */

fun GeoAddress(): GeoAddress = GeoAddress(Locale.getDefault())

val GeoAddress.street: String?
    get() {
        return subThoroughfare?.let { x ->
            thoroughfare?.let { y ->
                "$x $y"
            }
        }
    }

val GeoAddress.oneLineShort: String
    get() = oneLine()

fun GeoAddress.oneLine(withCountry: Boolean = false, withZip: Boolean = false): String = buildString {
    street?.let {
        append(it)
    }
    this@oneLine.locality?.let {
        append(' ')
        append(it)
    }
    this@oneLine.adminArea?.let {
        append(", ")
        append(it)
    }
    if (withCountry) {
        this@oneLine.adminArea?.let {
            append(' ')
            append(it)
        }
    }
    if (withZip) {
        this@oneLine.postalCode?.let {
            append(' ')
            append(it)
        }
    }
}.trim()

val GeoAddress.coordinate: GeoCoordinate? get() = if (hasLatitude()) GeoCoordinate(latitude, longitude) else null

File: ios/Khrysalis/com/lightningkite/khrysalis/location/GeoAddress.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation
import CoreLocation

//--- GeoAddress

public class GeoAddress {
    
    public let mark: CLPlacemark?
    public init(_ mark: CLPlacemark) {
        self.mark = mark
    }
    
    //--- GeoAddress()
    public init() {
        self.mark = nil
    }

    //--- GeoAddress.street
    public var street: String? {
        if let x = mark?.subThoroughfare, let y = mark?.thoroughfare {
            return "\(x) \(y)"
        }
        return nil
    }

    //--- GeoAddress.adminArea
    public var adminArea: String? {
        return mark?.administrativeArea
    }
    
    public var subAdminArea: String? {
        return mark?.subAdministrativeArea
    }

    //--- GeoAddress.locality
    public var locality: String? {
        return mark?.locality
    }
    
    public var subLocality: String? {
        return mark?.subLocality
    }

    //--- GeoAddress.countryName
    public var countryName: String? {
        return mark?.country
    }

    //--- GeoAddress.postalCode
    public var postalCode: String? {
        return mark?.postalCode
    }

    //--- GeoAddress.coordinate
    public var coordinate: GeoCoordinate? {
        return self.mark?.location?.coordinate.toKhrysalis()
    }

    public var oneLineShort: String {
        return oneLine()
    }

    //--- GeoAddress.oneLine(Boolean, Boolean)
    public func oneLine(_ withCountry: Bool, _ withZip: Bool = false) -> String {
        return oneLine(withCountry: withCountry, withZip: withZip)
    }
    public func oneLine(withCountry: Bool = false, withZip: Bool = false) -> String {
        var result = self.street ?? ""
        if let x: String = self.mark?.locality, let y: String = self.mark?.administrativeArea {
            result += " "
            result += x
            result += ", "
            result += y
        }
        if withCountry, let x = self.mark?.country {
            result += " "
            result += x
        }
        if withZip, let x = self.mark?.postalCode {
            result += " "
            result += x
        }
        return result.trim()
    }
}
